### PR TITLE
Apply microsoft code style format

### PR DESCRIPTION
**Pull request by shybovycha**
#### Description - What's this PR do?

Replace `clang-format` [settings](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#configurable-format-style-options) with a format derived from [Microsoft code style guidelines](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/code-style-rule-options?view=vs-2017).

Important bit: replaced the `style` option value with a valid one (`c++17`, not `Cpp11`).

#### Impacts/Risks of These Changes?

Code style only. Applied automatically with CLion.

#### Where should a reviewer start?

`.clang-format` or any source file (`*.hpp` or `*.cpp`). It is really just the spaces and blank lines.

#### How should this be tested?

n/a

#### Any background context you want to provide?

[A comment from FB](https://www.facebook.com/groups/cppEnthusiasts/permalink/3496258213766807?comment_id=3497222303670398&reply_comment_id=3497748800284415):

> Artem Shubovych
Sorry that the code format caused confusion. I imagine others could have the same problem with it. Can you suggest any changes to the formatting which would have helped you avoid this slipping from sight? I can think of a few things that might help but they all have trade offs. Were you reading that code from a web browser pointing at the github page for it? I ask because I'm wondering now if perhaps github itself provides maybe alternative coloring formats for C++ code that maybe would be easier to read.

#### Related Issues
List related issues.

#### Related Pull Requests
List related pull requests (if any).

### CODE REVIEW

File: PlayRho/Collision/AABB.cpp [hunk 1]
```diff
@@ -20,109 +20,111 @@
...
-{
-    auto sum = AABB{};
-    const auto xf = GetTransformation(world, id);
-    for (const auto& f: GetFixtures(world, id))
-    {
-        Include(sum, ComputeAABB(GetShape(world, f), xf));
-    }
-    return sum;
-}
+        AABB ComputeAABB(const World& world, const Body& body)
```
Comment (shybovycha) [inline PlayRho/Collision/AABB.cpp:80]: note how this is not really a change in the code, but just the method reordering. see lines 91..100

**Comment by louis-langholtz** [comment]
@shybovycha Thank you for your effort towards improving code readability by changing its formatting!

I took a look at these changes. Seems like one of the things it changes in the formatting is adding a level of indentation for each nested namespace. Is that correct?

On a side note, I don't see yet how changing the indentation for namespaces would consistently help with the readability problem of distinguishing free functions from member functions when those free functions would be in the same namespace (which they need to be to help with ADL). OTOH, I'd love to have more conversation about the code formatting in general!! 😃 

I'd actually started with a formatting style that indented declarations that are within namespaces and have since found myself disliking that and switching to not indenting declarations that are within namespaces. I love the use of namespaces but don't want to see code that's say 3 namespaces deep starting out with 4 levels of indentation. Here's an example of what I mean by that:

Instead of this indentation:
```
namespace a {
    namespace b {
        namespace c {

            inline int myfunction(float a) {
                return (a > 0.0f)? 1: (a < 0.0f)? -1: 0;
            }

        } // namespace c
    } // namespace b
} // namespace a
```

I'd prefer this indentation:
```
namespace a {
namespace b {
namespace c {

inline int myfunction(float a) {
    return (a > 0.0f)? 1: (a < 0.0f)? -1: 0;
}

} // namespace c
} // namespace b
} // namespace a
```

I'm not sure now which style the existing [`.clang-format`](https://github.com/louis-langholtz/PlayRho/blob/master/.clang-format) file uses for namespaces, but if it's the former, my apology. I'd prefer that clang format would format code using the latter, non-indented, style.

Some reasons why I prefer the non-indented style are as follows:
1. The non-indented style can be switched to C++17 nested namespace syntax without changing the indentation of any code within it thus minimizing changes in diffs if/when namespaces or their hierarchies change.
1. Code within say function definitions doesn't wind up starting already 4 levels of indentation deep and making code within its blocks more likely to run out of space before the end of the line biasing development of code towards higher levels of the namespace. I want functions in namespace `a::b::c` to have just as much space as any functions in the global namespace would have.

So changing the namespacing to using nested namespace syntax just produces the following:
```
namespace a::b::c {

inline int myfunction(float a) {
    return (a > 0.0f)? 1: (a < 0.0f)? -1: 0;
}

} // namespace a::b::c
```

And diffs don't show any changes to `myfunction`.

**Minimizing diff lines is a most compelling and objective of reasons for me for what I'd like the configuration of the `.clang-format` file to be.**

The changes you've suggested go beyond just the indentation (or not) of namespaces. Looks like enforcing an ordering of declarations is another one. I'd prefer to discuss each of those changes more individually if possible with examples preferably and then apply those changes that I'd like  to the code base's formatting. No small task I realize.

How about we start with creating markdown documentation on how clang format can be used on more targeted/common platforms (Windows, Linux, MacOS X) and/or particular IDEs (like MS Visual Studio, Eclipse, XCode). I've mostly just used `clang-format` from within MS Visual Studio and while I provided a `.clang-format` file in this repo, I don't myself use it when not on Windows.

**Comment by shybovycha** [comment]
## General discussion

### `clang-format` with specific platforms / IDEs

> how clang format can be used on more targeted/common platforms (Windows, Linux, MacOS X) and/or particular IDEs (like MS Visual Studio, Eclipse, XCode)

this is a very personalized thing, IMO. The way developer sets up their tools is not uniform and aiming for that would be an overkill, just as with providing project files for _some_ IDEs and not others. The thing I learned the hard way (working for a big company) is: no matter what you change - there will be people upset about that. The idea is to provide a config (like `.clang-format` or `CMakeLists.txt`) and then people will use it the way they want.

I am not entirely sure what do you mean by "how `clang-format` can be used on more targeted platforms and IDEs" - if you are talking specifically about how to configure menus and keyboard shortcuts for `clang-format` in IDEs? since the tool is supposed to work same no matter what platform it is running on.

The above two points are the reason for [my other PR](#358), removing all mentions of specific ways to build the project and leaving only CMake. No matter what platform you are working on, CMake shall run the same way and consume the same config. Depending on the correctness of its configuration, the generated build files (ninja, makefile, VS solution, etc.) might even build the thing correctly 😉  But that is slightly out of scope for this conversation 😛 

## Code format

### Nested namespaces

I was not sure if nested namespaces syntax is a thing in C++ yet, but seems like it is in C++20. I would prefer that one if possible (`namespace a::b::c`). Are you alright changing the requirements for the project to C++20-compatible compiler thought?

I do agree that indenting nested namespaces might make changes larger, but generally I don't think it is a bad idea - no indentation brings a space for misuse, whilst indenting every *code block* reduces that:

```
namespace a {

namespace b {

void  bar();

namespace c {

void moo();

}

void foo();

}

}
```

I bet this is possible and, although this is a very synthetic and rafinated example, imagine having 100x more code in between the lines. Could you tell which namespace each of the functions belongs to? Compare this to:

```
namespace a {

  namespace b {

    void  bar();

    namespace c {

      void moo();

    }

    void foo();

  }

}
```

Even with 2-space indentation this already solves the problem. Having guidelines in the editor improves that. Having the path to member (like in VS or CLion) goes even further.

### `.clang-format` settings

Let me do it step-by-step. I think best approach would be to open a test PR for that specific purpose, so here we are: #360

#### Getting prepped

[Commit](https://github.com/louis-langholtz/PlayRho/pull/360/commits/35458607f1646d24c9b41add48aff7d7f07def5e)

The first and foremost change is to remove everything from `.clang-format` and leave only the essential few settings:

```
Language: Cpp
BasedOnStyle: Microsoft
Standard: c++17
UseTab: Never
```

This brings some changes to the source code already:

```
namespace a {
namespace b {
}
}
```

becomes 

```
namespace a
{
namespace b
{
}
}
```

I think this is for good, as it keeps the braces positions consistent across the project (you are using braces on new line for methods already).

This also reorders imports:

```
#include <PlayRho/Collision/AABB.hpp>
#include <PlayRho/Collision/RayCastInput.hpp>
#include <PlayRho/Collision/DistanceProxy.hpp>
```

becomes

```
#include <PlayRho/Collision/AABB.hpp>
#include <PlayRho/Collision/DistanceProxy.hpp>
#include <PlayRho/Collision/RayCastInput.hpp>
```

I don't think this is a bad habit, but I do not have a super-strong opinion on that.

Removing trailing spaces is always a good thing IMO:

```diff
-    
+
```

It also reorders the members and aligns the comments

```
NullState, ///< point does not exist
AddState, ///< point was added in the update
```

becomes

```
AddState,     ///< point was added in the update
NullState,    ///< point does not exist
```

I do not think this is a bad thing.

Now, to the things that I probably want to change:

method params are joined into single line up to the wapping point:

```
DistanceOutput Distance(const DistanceProxy& proxyA, const Transformation& transformA,
                        const DistanceProxy& proxyB, const Transformation& transformB,
                        DistanceConf conf)
```

becomes

```
DistanceOutput Distance(const DistanceProxy &proxyA, const Transformation &transformA, const DistanceProxy &proxyB,
                        const Transformation &transformB, DistanceConf conf)
```

I personally prefer one param per line beyond some point:

```
DistanceOutput Distance(
    const DistanceProxy &proxyA, 
    const Transformation &transformA, 
    const DistanceProxy &proxyB,
    const Transformation &transformB,
    DistanceConf conf)
```

keen to hear your opinion on that.

The alignment of `switch..case` is changed:

```
    switch (count)
    {
        case 0:
            return false;
        case 1:
        {
            const auto v0 = proxy.GetVertex(0);
            const auto delta = point - v0;
            return GetMagnitudeSquared(delta) <= Square(vr);
        }
        default:
            break;
    }
```

becomes

```
    switch (count)
    {
    case 0:
        return false;
    case 1: {
        const auto v0 = proxy.GetVertex(0);
        const auto delta = point - v0;
        return GetMagnitudeSquared(delta) <= Square(vr);
    }
    default:
        break;
    }
```

I personally dislike that, but I am open to suggestions 😄 

Also, my one personalization to that is to have a blank line after each `case` block for better separation:

```
   switch (count)
    {
        case 0:
            return false;

        case 1:
        {
            const auto v0 = proxy.GetVertex(0);
            const auto delta = point - v0;
            return GetMagnitudeSquared(delta) <= Square(vr);
        }

        default:
            break;
    }
```

Empty method bodies are expanded to new lines:

```
     constexpr VariantData(UnusedData value) noexcept: unused{value} {}
```

becomes

```
    constexpr VariantData(UnusedData value) noexcept : unused{value}
    {
    }
```

I might live with that, but my personal preference is to keep these short (on the same line).

Template declarations are inlined:

```
template <std::size_t N>
struct LengthIndices
```

becomes

```
template <std::size_t N> struct LengthIndices
```

I personally dislike that.

Class members' access modifiers are indented:

```
class EdgeShapeConf : public ShapeBuilder<EdgeShapeConf>
{
public:
```

becomes

```
class EdgeShapeConf : public ShapeBuilder<EdgeShapeConf>
{
  public:
```

I also dislike that, but conditionally: this breaks the consistency with the aforementioned `switch..case` non-indented `case` statements. I would prefer either for both to be indented or none.

It does change the variable type annotations (like the ampersands and asterisks for references and pointers):

```diff
- AABB ComputeAABB(const DistanceProxy& proxy, const Transformation& xf) noexcept
+ AABB ComputeAABB(const DistanceProxy &proxy, const Transformation &xf) noexcept
```

This specific example is fine, but then there are the uses of these types in, say, templates:

```diff
- return static_cast<ConcreteConf&>(*this);
+ return static_cast<ConcreteConf &>(*this);
```

I personally dislike this last bit.

That's for the default code style setup. Can go deeper into examples of some other options, but I highly recommend that you try them out yourself in some IDE like CLion (it has the preview).

**Comment by louis-langholtz** [comment]
> not sure if nested namespaces syntax is a thing in C++ yet, but seems like it is in C++20

Nested namespaces made it into [C++17](https://en.wikipedia.org/wiki/C%2B%2B17).

C++20 has some interesting changes but I don't intend to update this project's standard to it any time soon. Seems compilers like Clang/LLVM, gcc, and MSVC lag years behind the standard in terms of providing complete working implementations.

> The first and foremost change is to remove everything from .clang-format and leave only the essential few settings

It's disappointing to me that the ClangFormat development seems to be favoring just a few format guides via its `BasedOnStyle` parameter but it's understandable. I suspect they're doing that as a practical matter since more arbitrary formatting styles would seem to require more endless configuration choices.

In any case, selecting a specific value for `BasedOnStyle` does seem desirable to me. Ideally, it'd be the style that's closest to what the style in code is now. I tried searching the web for info on what the default `BasedOnStyle` value setting is when it's not stated in the `.clang-format` file but didn't find anything about that. Presumably a default is used but maybe not. Need to look at the source code for clang-format apparently.

> This brings some changes to the source code already.

The indentation style I prefer for namespaces is:
```
namespace a {
namespace b {
} // namespace b
} // namespace a
```

I don't want to go away from this style.

> This also reorders imports

Consistent ordering of imports is desirable to me. The latest `.clang-format` doesn't establish any order I don't believe. Some includes will need to be manually arranged however so that sorting only happens within adjacent lines (those that aren't separated by an empty line).

> Removing trailing spaces is always a good thing IMO

Agreed. 😄 

> Class members' access modifiers are indented... I also dislike that

Agreed. 😄 

I won't be accepting a change to the formatting that indents member access modifiers. I'd sooner want to modify ClangFormat to provide support for customization of this. 

> It does change the variable type annotations (like the ampersands and asterisks for references and pointers)

That's a debate unto itself much like [East const vs. West const](https://stackoverflow.com/questions/5503352/const-before-or-const-after).

As I [wrote in the other PR/Discussion on this](https://github.com/louis-langholtz/PlayRho/pull/360):

> **Admittedly, I care more about things like the correctness of the code, its performance, and its documentation, than about the style in which it's written.**

**Comment by louis-langholtz** [comment]
Embarrassingly perhaps, it took me a while to figure out that, if left unspecified, `BasedOnStyle` defaults to `LLVM`. In commit e82631b35bad782e22ee8e2b25a1abe861bd5cbd, I went ahead and added that explicitly to the `.clang-format` file so that won't be as hard for me recognize in the future.

Anyways, while the project has a `.clang-format` file, ClangFormat has never been run over every file in the project. It's intended more as a formatting tool for any new files. I know of at least one IDE that doesn't have integrated support for ClangFormat (Xcode) and that partly discourages me from making ClangFormat more of a requirement.

If you'd like to make a PR for just changing the `.clang-format` file from what it is now (that doesn't include changes to any other files) with explanation of what the changes will help with and why, that'd be preferable to me over considering a PR that changes 200+ files.

**Pull request closed without merging**